compiler/rustc_middle/src/ty/mod.rs

```diff
@@ -1501,33 +1511,44 @@ impl ReprOptions {
             }
         }
 
+        // If `-Z randomize-layout` is enabled then we can consider performing field randomization
+        if tcx.sess.opts.debugging_opts.randomize_layout {
+            flags.insert(ReprFlags::CAN_RANDOMIZE);
+        }
```

Comment:
I think the semantics are a big off. The flag *forces* randomization, no? So I'd maybe call the flag `FORCE_FIELD_SHUFFLING` / `FORCE_FIELD_RANDOMIZATION`.

This would also fit with the idea that *some* amount of (pseudo)randomization would always happen, and the flag only exists for doing it *at the expense of* size optimizations.

Reply:
> The flag forces randomization, no?

In `can_randomize_struct_fields` it checks `inhibit_struct_field_reordering_opt` in addition to the flag.

Reply:
The flag allows for the possibility of randomization, as the8472 said it still checks that the struct can actually have layout optimizations applied to it before giving the go-ahead

Reply:
One thing is that it should probably (if possible) break layout optimizations where it isn't guaranteed, though this would, of course, require introducing a lang item on `Option<T>`.

Reply:
I believe the `Option<T>` guarantees are for every type of the same shape as `Option<T>`, but I may be wrong.

Reply:
IIRC, it's only Option specifically (in the language itself), though it could be reasonable to implement it so that it works for the guaranteed niche-optimized types, reguardless of the enum.